janus_server/src/datastore.rs

```diff
@@ -3812,6 +3932,475 @@ mod tests {
+            state: AggregationJobState::Finished,
+        }];
+        let report_aggregations = vec![ReportAggregation::<FakeVdaf> {
+            aggregation_job_id,
+            task_id,
+            nonce: reports[0].nonce(),
+            ord: 0,
+            // Shouldn't matter what state the report aggregation is in
+            state: ReportAggregationState::Start,
+        }];
+
+        let collect_job_test_cases = vec![CollectJobTestCase {
+            should_be_acquired: false,
+            task_id,
+            batch_interval: Interval::new(
+                Time::from_seconds_since_epoch(0),
+                Duration::from_seconds(100),
+            )
+            .unwrap(),
+            agg_param: 0u8,
+            collect_job_id: None,
+            set_helper_aggregate_share: false,
+        }];
+
+        run_collect_job_acquire_test_case(
+            &ds,
+            CollectJobAcquireTestCase {
+                task_ids: vec![task_id],
+                reports,
+                aggregation_jobs,
+                report_aggregations,
+                collect_job_test_cases,
+            },
+        )
+        .await;
+    }
+
+    #[tokio::test]
+    async fn collect_job_acquire_release_job_finished() {
+        install_test_trace_subscriber();
+        let clock = MockClock::default();
+        let (ds, _db_handle) = ephemeral_datastore(clock.clone()).await;
+
+        let task_id = TaskId::random();
+        let reports = vec![new_dummy_report(task_id, Time::from_seconds_since_epoch(0))];
+        let aggregation_job_id = AggregationJobId::random();
+        let aggregation_jobs = vec![AggregationJob::<FakeVdaf> {
+            aggregation_job_id,
+            aggregation_param: 0u8,
+            task_id,
+            state: AggregationJobState::Finished,
+        }];
+        let report_aggregations = vec![ReportAggregation::<FakeVdaf> {
+            aggregation_job_id,
+            task_id,
+            nonce: reports[0].nonce(),
+            ord: 0,
+            state: ReportAggregationState::Start,
+        }];
+
+        let collect_job_test_cases = vec![CollectJobTestCase {
+            should_be_acquired: false,
+            task_id,
+            batch_interval: Interval::new(
+                Time::from_seconds_since_epoch(0),
+                Duration::from_seconds(100),
+            )
+            .unwrap(),
+            agg_param: 0u8,
+            collect_job_id: None,
+            // Collect job has already run to completion
+            set_helper_aggregate_share: true,
+        }];
+
+        run_collect_job_acquire_test_case(
+            &ds,
+            CollectJobAcquireTestCase {
+                task_ids: vec![task_id],
+                reports,
+                aggregation_jobs,
+                report_aggregations,
+                collect_job_test_cases,
+            },
+        )
+        .await;
+    }
+
+    #[tokio::test]
+    async fn collect_job_acquire_release_aggregation_job_in_progress() {
```

Comment:
This test fails because as I described elsewhere, I can't quite figure out the SQL for it.